add premium_available_in_country context var

groovecoder opened the pull request:
* Adds a new `/api/v1/premium_countries` endpoint with response like:
```
{
"country_code":"us",
"premium_countries":["at","be","ch","de","es","fr","ie","it","nl","us","gb","ca","nz","my","sg"],
"premium_available_in_country":true
}
```
* Also adds the same 3 items as context vars (`country_code`, `premium_countries`, and `premium_available_in_country`) for all templates
* Moves the common code for those 3 items into a new `privaterelay.utils.get_premium_countries_info_from_request` function.

Vinnl commented on privaterelay/context_processors.py:
```diff
@@ -12,13 +12,17 @@ def common(request):
     avatar = fxa.extra_data['avatar'] if fxa else None
     accept_language = request.headers.get('Accept-Language', 'en-US')
-    country_code = request.headers.get('X-Client-Region', 'us').lower()
+    country_code = request.headers.get('X-Client-Region', '').lower()
```
I'd do one of two things here:

1. Fall back to `accept_language` if this header is not available (risking that we're showing the wrong price/currency to people, and/or showing Premium promotion to people who can't get it), or
2. Set `country_code` to `us` if the header is not set and `DEBUG==True`, otherwise we'll never see Premium banners when running locally, possibly breaking it without noticing.

groovecoder replied:
Moved a lot of this into `utils.py` now too.

Vinnl replied:
Looks good. I guess the second point isn't too relevant if you fall back to the Accept language too, since the browser I think always sends an accept language and thus we'll never hit that code path in practice, but it doesn't do any harm either :)

Vinnl commented on privaterelay/utils.py:
```diff
@@ -0,0 +1,37 @@
+    if 'X-Client-Region' in request.headers:
+        return request.headers['X-Client-Region'].lower()
+    if 'Accept-Language' in request.headers:
+        return get_premium_country_lang(request.headers['Accept-Language'])
```
As I understand it `get_premium_country_lang` returns a tuple `(cc, lang)`. If we don't return just the country code, `get_premium_countries_info_from_request` will try to find the tuple as a whole in `premium_countries`, which it can't.

```suggestion
        return get_premium_country_lang(request.headers['Accept-Language'])[0]
```

maxxcrawford replied:
@groovecoder This change is blocking #1376. Could you review?

groovecoder replied:
Committed suggestion.

Vinnl commented on privaterelay/utils.py:
```diff
@@ -0,0 +1,37 @@
+        if lang in languages.keys():
+            return cc, lang
+        return cc, list(languages.keys())[0]
+    return 'us', 'en'
```
This causes a problem for detecting whether a country has Premium based on the accept language: if the country does not have Premium, this will auto-default to `'us', 'en'`, and then return that they _do_ have Premium.

say-yawn replied:
@Vinnl said good catch! @groovecoder, we should return the `cc` as is and catch and handle that in `premium_plan_id` and `premium_plan_price`.

groovecoder replied:
Fixed and added test!

say-yawn commented on privaterelay/utils.py:
```diff
@@ -0,0 +1,37 @@
+        return request.headers['X-Client-Region'].lower()
+    if 'Accept-Language' in request.headers:
+        return get_premium_country_lang(request.headers['Accept-Language'])[0]
+    if settings.DEBUG:
+        return 'us'
```
```suggestion
```
Is this test code or actually necessary?

groovecoder replied:
We want this so that our local dev instances are "premium available" by default.

say-yawn reviewed (approved):
LGTM. Minor change recommended.

Vinnl commented on privaterelay/utils.py:
```diff
@@ -0,0 +1,37 @@
+        if lang in languages.keys():
+            return cc, lang
+        return cc, list(languages.keys())[0]
+    return cc, 'en'
```
@groovecoder Hmm, looks like we have other code that assumes that this function only returns country codes that are present in `PREMIUM_PLAN_COUNTRY_MAPPING`, e.g. here: https://github.com/mozilla/fx-private-relay/blob/68c209d2c447c5fbe69a2c4216c8685bccc51b8d/privaterelay/templatetags/relay_tags.py#L57

If I run this and set my browser's Accept-Language to `ro-ro`, I get a KeyError because `ro` is not present there :/

groovecoder replied:
argh!!! okay let me add a test case and fix.

groovecoder replied:
Okay, I updated the `relay_tags` to return sensible values when the country code isn't in the `MAPPING`. (I think I checked that all places that need the ID will work with a `''` value, and everywhere that needs a price can at least *render* a USD price instead.

Vinnl replied:
Testing it locally, I'm no longer encountering any problems! :tada:

Vinnl commented:
Oh, could I make one more request for this API? It would be nice if it could also return the plan ID for the Premium plan relevant to this country.

groovecoder commented:
I rebased this branch on `main` which caused conflicts and broke comparison in GitHub, so I didn't merge #1396. But, I added `plan_country_lang_mapping` to `/api/v1/premium_countries` in this branch myself.
